File: examples/android-kotlin/app/src/main/java/com/example/waku/messages/Message.kt
package com.example.waku.messages

import com.example.waku.handleResponse
import com.example.waku.serializers.ByteArrayBase64Serializer
import gowaku.Gowaku
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json

@Serializable
data class Message(
    @Serializable(with = ByteArrayBase64Serializer::class) val payload: ByteArray,
    val contentTopic: String? = "",
    val version: Int? = 0,
    val timestamp: Long? = null
)

/**
 * Decode a waku message using an asymmetric key
 * @param msg Message to decode
 * @param privateKey Secp256k1 private key used to decode the message
 * @return DecodedPayload containing the decrypted message, padding, public key and signature (if available)
 */
fun Message.decodeAsymmetric(privateKey: String): DecodedPayload {
    val jsonMsg = Json.encodeToString(this)
    val response = Gowaku.decodeAsymmetric(jsonMsg, privateKey)
    return handleResponse<DecodedPayload>(response)
}

/**
 * Decode a waku message using a symmetric key
 * @param msg Message to decode
 * @param privateKey Symmetric key used to decode the message
 * @return DecodedPayload containing the decrypted message, padding, public key and signature (if available)
 */
fun Message.decodeSymmetric(symmetricKey: String): DecodedPayload {
    val jsonMsg = Json.encodeToString(this)
    val response = Gowaku.decodeSymmetric(jsonMsg, symmetricKey)
    return handleResponse<DecodedPayload>(response)
}

File: examples/swift-waku/swift-waku/Message.swift
//
//  Message.swift
//  swift-waku
//

import Foundation
import Gowaku

class Message: Codable {
    var payload: [UInt8] = [];
    var contentTopic: String? = "";
    var version: Int? = 0;
    var timestamp: Int64? = nil;
    
    func decodeAsymmetric(_ privateKey: String) throws -> DecodedPayload {
        let jsonEncoder = JSONEncoder()
        let jsonData = try! jsonEncoder.encode(self)
        let jsonMsg = String(data: jsonData, encoding: String.Encoding.utf8)
        let response = GowakuDecodeAsymmetric(jsonMsg, privateKey)
        return try handleResponse(response)
    }

    func decodeSymmetric(symmetricKey: String) throws -> DecodedPayload {
        let jsonEncoder = JSONEncoder()
        let jsonData = try! jsonEncoder.encode(self)
        let jsonMsg = String(data: jsonData, encoding: String.Encoding.utf8)
        let response = GowakuDecodeSymmetric(jsonMsg, symmetricKey)
        return try handleResponse(response)
    }
}
